cheatsheets_draft/GraphQL_Cheat_Sheet.md

```diff
@@ -0,0 +1,257 @@
+
+Many implementations of GraphQL have Introspection and GraphiQL enabled by default and leave them accessible without requiring authentication. This is problematic because introspection allows the requester to learn all about supported schema and queries (see a [real-world example](https://hackerone.com/reports/291531) abusing this). Introspection might be how the API owner wants to educate consumers about how to use the API. However, the preferred way to educate consumers about a service is through a separate documentation channel such as a wiki, Git Readme, or readthedocs.
+
+The safest and usually easiest approach is to just disable introspection and GraphiQL system-wide. See [this page](https://lab.wallarm.com/why-and-how-to-disable-introspection-query-for-graphql-apis/) or consult your GraphQL implementation's documentation to learn how to disable introspection altogether. If your implementation does not natively support disabling introspection or if you would like to allow some consumers/roles to have this access you can build a filter in your service to only allow approved consumers to access the introspection system.
+
+> This blurb is from nikitastupin. Would be great if we knew the name of this "hint" feature and link to documentation to disable it. I couldn't find anything from a quick Google.
+
+Keep in mind that even if introspection is disabled you can still guess fields by brute forcing them. Furthermore, some GraphQL implementations give you a hint when a field name you provide is similar to an existing field (e.g. you provide `usr` and the response may ask if you meant the valid `user` field instead). You should consider disabling this feature.
+
+***Disable Introspection - Java***
+
+```Java
+GraphQLSchema schema = GraphQLSchema.newSchema()
+    .query(StarWarsSchema.queryType)
+    .fieldVisibility( NoIntrospectionGraphqlFieldVisibility.NO_INTROSPECTION_FIELD_VISIBILITY )
+    .build();
+```
+
+***Disable Introspection & GraphiQL - JavaScript***
+
+```javascript
+app.use('/graphql', graphqlHTTP({
+  schema: MySessionAwareGraphQLSchema,
++ validationRules: [NoIntrospection]
+  graphiql: process.env.NODE_ENV === 'development',
+}));
+```
+
+### IDOR Protection
+
+In general, proper access controls will prevent any IDOR attacks since IDOR merely represents 2 separate authorization issues: [Broken Object Level Authorization](https://github.com/OWASP/API-Security/blob/master/2019/en/src/0xa1-broken-object-level-authorization.md) and [Broken Function Level Authorization](https://github.com/OWASP/API-Security/blob/master/2019/en/src/0xa5-broken-function-level-authorization.md). See the [IDOR prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Insecure_Direct_Object_Reference_Prevention_Cheat_Sheet.html) for a comprehensive rundown of preventing IDOR.
+
+> New section from nikitastupin. Not something I can personally verify with my limited GraphQL experience, so would be great if someone else can check this out.
```

Comment:
You can check GitHub's GraphQL schema https://api.github.com/graphql, they have such a field.

Reply:
If I do end up removing the IDOR section I would probably leave this section out too. Maybe I could add the below note to the "Access Control" section in place of this section:

> Be careful with the `node` or `nodes` fields in query objects because they can be used to access objects directly by `ID`. Part of proper authorization checks includes ensuring that the requester has access to the objects they are requesting by ID. The requester supplying the ID of an object is not sufficient to prove they have access to it.

Maybe I will also include the command to check if the schema has these fields.

Thoughts on that?

Reply:
I'm trying to find some notes in the documentation about how to disable the node or nodes functionality, but it seems like the docs say all GraphQL APIs must support this: https://graphql.org/learn/global-object-identification/. Do you have any documentation or examples that show disabling the node or nodes field?